### PR TITLE
docstring and defaults

### PR DESCRIPTION
**Pull request by ocefpaf**
Addressing some of the points we raised in https://github.com/python-visualization/folium/pull/187.

@andrewgiessel since this is your baby can you please take a look?

Here is a working example:
http://nbviewer.ipython.org/github/ocefpaf/folium_notebooks/blob/master/test_image_overlay_gulf_stream.ipynb

### CODE REVIEW

**Comment by andrewgiessel** [comment]
This looks great!  Thanks for tackling it.

A comment:  While I've never really been satisfied with `data_projection` as an argument name, I do like the explicitness of using 'mercator' and None.  That said, I don't think we should incorporate arbitrary projections (like using proj4 or whatever, blech)  If we use True/False, then maybe 'project_to_mercator' might be a clearer argument name.  There is just a lot of ambiguity.

Also, projection only works if you use it with an array, not with a local or external file.

**Comment by ocefpaf** [comment]
I am not happy with `data_projection` either.  To be honest I am not happy with that option
and the current implementation.  The image distortion seen [here](http://nbviewer.ipython.org/github/ocefpaf/folium_notebooks/blob/master/test_image_overlay.ipynb) worries me...

Maybe should remove that for now and try to implement that using [rasterio](https://github.com/mapbox/rasterio) or Cartopy.

PS: Any idea about what is going on with travis?

**Comment by andrewgiessel** [comment]
Ah, I think the issue is that the algorithm is for global data, not local.
So, for my global data overlays it works well.  For smaller single images,
it won't.  So, generally speaking, I think we 2 have different use cases
(which are both very important):  user generated data (my primary use), and
landsat data (yours?).  The former will tend to be global data, and this
algo will work.  The later doesn't work well.

rasterio looks pretty cool.  I don't know enough about cartopy to know how
to do this, but I should learn about both- do you want me to do that and
then look into a more general solution?

**Comment by ocefpaf** [comment]
I am re-thinking this issue and  I believe it is better to avoid any re-projection inside folium.  Lets just say it takes the image in `Web Mercator` and leave the transformation to the user.  My notebook shows one example of re-projecting Geodetic data to PlateCarree (which seems to be OK in the map.)

We can easily create example with your `geodetic_to_mercator()` and `rasterio` too.  But I am leaning towards not hard-wiring this into folium.

**Comment by andrewgiessel** [comment]
Ya, projection is obviously a difficult problem and generally outside the
scope of folium, I agree.

On the other hand, the utility of not having to use cartopy, make a figure,
and save a temporary file for simple overlays of numpy arrays is pretty
huge for me (and others, I believe).  The likelihood is high that anything
generated in numpy will be geodetic, so some projection needs to be done.

If `geodetic_to_mercator` is left in folium.utilities (and perhaps
referenced in the docstrings or docs?) I think this is an OK compromise.  A
globally dataset is a common use case for generated data, and we account
for that.  Anything else can be handled by the user.

ag

**Comment by ocefpaf** [comment]
> If `geodetic_to_mercator` is left in folium.utilities (and perhaps
> referenced in the docstrings or docs?) I think this is an OK compromise.

That makes sense. As soon as get back in the office I will change this PR
to include the transformation back into utilities.  We just need an example
with globaly generated data for the docstring and I think I got one.

**Comment by ocefpaf** [comment]
We can try to use this in the global exemple:

http://scitools.org.uk/cartopy/docs/latest/examples/waves.html
Em 18/08/2015 11:42, "Filipe Pires Alvarenga Fernandes" <ocefpaf@gmail.com>
escreveu:

> > If `geodetic_to_mercator` is left in folium.utilities (and perhaps
> > referenced in the docstrings or docs?) I think this is an OK compromise.
>
> That makes sense. As soon as get back in the office I will change this PR
> to include the transformation back into utilities.  We just need an example
> with globaly generated data for the docstring and I think I got one.
>

**Comment by andrewgiessel** [comment]
Ya, good find.  That example works well:

image

image


The latter is correct.

**Comment by andrewgiessel** [comment]
You can wrap `data` in a MPL colormap for fancy colors.

**Comment by ocefpaf** [comment]
OK.  I am done with this PR.  [Here](http://nbviewer.ipython.org/github/ocefpaf/folium_notebooks/blob/master/test_image_overlay_geodetic.ipynb) is an example that needs the projection step.  The same example is illustrated in the docstrings.

Summary:
- Renamed the `opacity` option to `image_opacity` to be consistent with the rest of the API;
- The `image_opacity` is no transparency (`1`);
- No automatic projection is performed.  The method assumes the image/data is in Web Mercator.
- There is an example and a utility function to project the data as Web Mercator (`folium.utilities.geodetic_to_mercator`)

@andrewgiessel can you take another look and give the :+1:  or :-1: 

After that @BibMartin or @themiurgo can take another look and merge.

File: folium/folium.py [hunk 1]
```diff
@@ -951,44 +951,37 @@ def json_style(style_cnt, line_color, line_weight, line_opacity,
...
+        """
+        Simple image overlay of raster `data` from an image or a NumPy array.
+        This is a lightweight way to overlay geospatial data on top of a map.
+        The default lon, lat corners are the whole globe.  Adjust it to your
+        image.
+
+        If `data` is is a NumPy array a PNG file will be created for you.
+        If you do not specify a filename the image will be embedded inline.
+
+        NOTE: The image must be project as Web Mercator
```
Comment (andrewgiessel) [inline folium/folium.py]: s/project/projected

File: folium/folium.py [hunk 2]
```diff
@@ -951,44 +951,37 @@ def json_style(style_cnt, line_color, line_weight, line_opacity,
...
-            if numpy array, must be a image format,
-            i.e., NxM (mono), NxMx3 (RGB), or NxMx4 (RGBA)
-            if url, must be a valid url to a image (local or external)
-        opacity: float, default 0.75
-            Image layer opacity in range 0 (transparent) to 1 (opaque)
-        min_lat: float, default -90.0
-        max_lat: float, default  90.0
-        min_lon: float, default -180.0
-        max_lon: float, default  180.0
+        data: NumPy array OR image path.
```
Comment (andrewgiessel) [inline folium/folium.py]: or external URL

**Comment by andrewgiessel** [comment]
Some minor comments in the docstring, but :+1: from me!  Great work and thanks for polishing this up so that it's more consistent with the whole project.

File: folium/folium.py (hunk 2 above)
Comment (ocefpaf) [inline folium/folium.py]: Virtually all the external data file (GeoJSONs, images) can be local or a URL.  To keep it consistent with the other docstrings I will leave just `image` ~~path~~.

File: folium/folium.py (hunk 1 above)
Comment (ocefpaf) [inline folium/folium.py]: Thanks!

**Comment by ocefpaf** [comment]
One thing we need to come back when we pluginify this is the inline vs path reference.  I think this should be a global option (standalone HTML vs separate files) and not inside the method.

**Comment by ocefpaf** [comment]
@BibMartin we have @andrewgiessel :+1: and green `Travis`.  Can you take a quick look and merge?

**Comment by BibMartin** [comment]
It seems good. Congrats and thanks to both of you!

**Pull request merged**